File: LC 295.cpp
class MedianFinder {
    priority_queue<int> pq1;
    priority_queue<int, vector<int>, greater<int>> pq2;
public:
    MedianFinder() {

    }
    
    void addNum(int num) {
        if (pq1.empty() || num < pq1.top() ) {
            pq1.push(num);
        } else {
            pq2.push(num);
        }
        while (pq1.size() > pq2.size() + 1) {
            auto t = pq1.top();
            pq1.pop();
            pq2.push(t);
        }
        while (pq1.size() < pq2.size()) {
            auto t = pq2.top();
            pq2.pop();
            pq1.push(t);
        }
    }
    
    double findMedian() {
        if (pq1.size() == pq2.size()) {
            return 1.0 * (pq1.top() + pq2.top()) / 2;
        } 
        return pq1.top();
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 1000/LC 295.go

type IntHeap []int
func (h IntHeap) Len() int {return len(h)}
func (h IntHeap) Less(i, j int) bool {return h[i] > h[j]}
func (h IntHeap) Swap(i, j int) {h[i], h[j] = h[j], h[i]}
func (h *IntHeap) Push(x any) {
    *h  = append(*h, x.(int))
}
func (h *IntHeap) Pop() any {
    old := *h
    n := len(old)
    x := old[n-1]
    *h = old[:n-1]
    return x
}

type MedianFinder struct {
    pq1 IntHeap
    pq2 IntHeap
}


func Constructor() MedianFinder {
    return MedianFinder{IntHeap{}, IntHeap{}}
}


func (this *MedianFinder) AddNum(num int)  {
    if this.pq1.Len() == 0 || this.pq1[0] >= num {
        heap.Push(&this.pq1, num)
    } else {
        heap.Push(&this.pq2, -num) 
    }
    
    for this.pq2.Len() + 1 < this.pq1.Len() {
        t := -heap.Pop(&this.pq1).(int)
        heap.Push(&this.pq2, t)
    }
    for this.pq1.Len() < this.pq2.Len() {
        t := -heap.Pop(&this.pq2).(int)
        heap.Push(&this.pq1, t)
    }
}


func (this *MedianFinder) FindMedian() float64 {
    if this.pq1.Len() > this.pq2.Len() {
        return float64(this.pq1[0])
    }
    return float64(this.pq1[0] - this.pq2[0]) / 2.0
}


/**
 * Your MedianFinder object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddNum(num);
 * param_2 := obj.FindMedian();
 */
